### PR TITLE
ENT-4515 rotate keys on upgrade

### PR DESCRIPTION
**Pull request by Lex-2008**

### CODE REVIEW

**Comment by cf-bottom** [comment]
Thank you for submitting a pull request! Maybe @craigcomstock can review this?

File: packaging/common/cfengine-hub/postinstall.sh [hunk 1]
```diff
@@ -971,6 +971,37 @@ else
     cf_console platform_service cfengine3 start
 fi
 
+if is_upgrade; then
+    set +x
+    pwgen() {
+	    dd if=/dev/urandom bs=1024 count=1 2>/dev/null | tr -dc 'a-zA-Z0-9' | fold -w $1 | head -n 1
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: weird indentation here

File: packaging/common/cfengine-hub/postinstall.sh [hunk 2]
```diff
@@ -971,6 +971,37 @@ else
...
+    }
+
+    pwhash() {
+	    echo -n "$1" | openssl dgst -sha256 | awk '{print $2}'
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: Also weird indentation.

File: packaging/common/cfengine-hub/postinstall.sh [hunk 3]
```diff
@@ -971,6 +971,37 @@ else
...
+    sed -i "/MP_CLIENT_SECRET/s/.*/\$config['MP_CLIENT_SECRET'] = '$MP_CLIENT_SECRET';/" $PREFIX/share/GUI/application/config/appsettings.php
+    sed -i "/LDAP_API_SERVER_SECRET/s/.*/\$config['LDAP_API_SERVER_SECRET'] = '$MP_CLIENT_SECRET';/" $PREFIX/share/GUI/application/config/appsettings.php
+    sed -i "/config..encryption_key/s/.*/\$config['encryption_key'] = '$MP_CLIENT_SECRET';/" $PREFIX/share/GUI/application/config/config.php
+    sed -i "/VALUES...MP/s/.*/INSERT INTO oauth_clients VALUES ('MP', '$MP_CLIENT_SECRET', '', 'password refresh_token', NULL, NULL);/" $PREFIX/share/db/ootb_settings.sql
+
+    if test -d $PREFIX/share/GUI/ldap; then # New LDAP back end introduced in 3.11.0
+	    sed -i "/LDAP_API_SERVER_SECRET/s/.*/\$config['LDAP_API_SERVER_SECRET'] = '$MP_CLIENT_SECRET';/" $PREFIX/share/GUI/application/config/appsettings.php
+	    sed -i "/accessToken/s/.*/    'accessToken' => '$MP_CLIENT_SECRET',/" $PREFIX/share/GUI/ldap/config/settings.php
+	    sed -i "/LDAP_API_SECRET_KEY/s/.*/define('LDAP_API_SECRET_KEY', '$MP_CLIENT_SECRET');/" $PREFIX/share/GUI/api/config/config.php
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: Again, really big indentation.

File: packaging/common/cfengine-hub/postinstall.sh [hunk 4]
```diff
@@ -971,6 +971,37 @@ else
...
+    MP_CLIENT_SECRET=`pwgen 32`
+    CFE_ROBOT_PW=`pwgen 32`
+    CFE_ROBOT_PW_SALT=`pwgen 10`
+    CFE_ROBOT_PW_HASH=`pwhash "$CFE_ROBOT_PW_SALT$CFE_ROBOT_PW"`
+
+    sed -i '/CFE_ROBOT_PASSWORD/s/.*/$config["CFE_ROBOT_PASSWORD"] = "'$CFE_ROBOT_PW'";/' $PREFIX/httpd/htdocs/application/config/cf_robot.php
+    sed -i "/MP_CLIENT_SECRET/s/.*/\$config['MP_CLIENT_SECRET'] = '$MP_CLIENT_SECRET';/" $PREFIX/share/GUI/application/config/appsettings.php
+    sed -i "/LDAP_API_SERVER_SECRET/s/.*/\$config['LDAP_API_SERVER_SECRET'] = '$MP_CLIENT_SECRET';/" $PREFIX/share/GUI/application/config/appsettings.php
+    sed -i "/config..encryption_key/s/.*/\$config['encryption_key'] = '$MP_CLIENT_SECRET';/" $PREFIX/share/GUI/application/config/config.php
+    sed -i "/VALUES...MP/s/.*/INSERT INTO oauth_clients VALUES ('MP', '$MP_CLIENT_SECRET', '', 'password refresh_token', NULL, NULL);/" $PREFIX/share/db/ootb_settings.sql
```
Comment (vpodzime) [inline packaging/common/cfengine-hub/postinstall.sh]: Why are the patterns more generic than the replacements? Shouldn't the patterns have all the brackets, the full `INSERT` statement and similar?

Comment (Lex-2008) [reply packaging/common/cfengine-hub/postinstall.sh]: I don't have a strong opinion here, my idea is that these files shouldn't be edited by customers, so we know what they contain, so patterns can be as small as we want - they just need to match a single line. In other words, if you have a file with two lines:

```
aabbbabab
cccddcdcd
```

and you need to match the second line - your pattern can be just a single `c`.

Comment (vpodzime) [reply packaging/common/cfengine-hub/postinstall.sh]: Yes, but at the same time they should really just change the specific lines, right? I agree the patterns here don't have to be crazy complex, but I can definitely see this braking something in the future. Like if somebody adds a line like `$config['encryption_key_length']` or similar.

Ideally, these secrets should be in separate file(s) that **only** contain the secrets and we should be (over)writing whole files in the scripts with something simple like `cat <<EOF`.

**Comment by Lex-2008** [comment]
Yup, created https://tracker.mender.io/browse/ENT-4594

Also I made match patterns to be more specific to help with cases like `$config['encryption_key_length']`. But ofc they will not save you if you put two statements in one line, like this:

```
$config['encryption_key'] = '…'; do_something_important(1024);
```

**Comment by Lex-2008** [comment]
good idea! @cf-bottom make me a build, please

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/2396/)

**Jenkins:** https://ci.cfengine.com/job/pr-pipeline/2396/

**Packages:** http://buildcache.cfengine.com/packages/testing-pr/jenkins-pr-pipeline-2396/

**Comment by Lex-2008** [comment]
Tested above build, keys rotated ok, merging

**Pull request merged**